File: CaseStudies/FileInputTests/Bundled/Design3/main.jl
# Bundled

#=
    This does seem like the 'right way' to do it for a procedural language.
    It feels off for OO though - validation should be kept within the class.
=#

# Struct for holding parameters.
struct Parameters
    name::AbstractString
    height::Integer
    weight::AbstractFloat
end

# We don't even need a constructor, so that simplifies things a lot

# No longer part of the Parameters 'class'
# Validation is done inside here now
function get_input(filename::AbstractString)

    file = open(filename, "r")
    lines = readlines(file)
    close(file)

    name = lines[2]
    if isempty(strip(name))
        throw("Name must not be blank!")
    end
    height = parse(Int32, lines[4])
    if height <= 0
        throw("Height must be positive!")
    end
    weight = parse(Float64, lines[6])
    if weight <= 0
        throw("Weight must be positive!")
    end

    return Parameters(name, height, weight)
end

p = get_input("FileInputTests/input.txt")
println(p)

File: CaseStudies/FileInputTests/Bundled/Design3/main.swift
// Bundled

/* 
    I would rather have validation inside the class, since
    validation is 'part of' the class, wheras parsing is more
    'a way of getting' the class
*/
import Foundation

// Class for holding parameters
class Parameters {
    let name: String // Default values not needed now!
    let height: Int
    let weight: Double

    // Very basic initializer now.  We could even use a struct instead 🤔
    init(_ name: String,_ height: Int, _ weight: Double) throws {
        (self.name, self.height, self.weight) = (name, height, weight)
    }

}

func get_input(_ filename: String) throws -> Parameters {

    let lines = try String(contentsOfFile: filename, encoding: String.Encoding.utf8)
        .components(separatedBy: "\n")
    
    let name: String = lines[1]
    if name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
        throw ParamError("Name must not be blank!")
    }    
    let height: Int = Int(lines[3]) ?? 0
    if height <= 0 {
        throw ParamError("Height must be positive!")
    }
    let weight: Double = Double(lines[5]) ?? 0
    if weight <= 0 {
        throw ParamError("Weight must be positive!")
    }

    return try Parameters(name, height, weight)
}

// Simple error struct
struct ParamError: LocalizedError {
    let description: String

    init(_ description: String) {
        self.description = description
    }

    var errorDescription: String? {
        description
    }
}

var p: Parameters = try get_input("FileInputTests/input.txt")
print("Parameters(\"" + p.name + "\", " + String(p.height) + ", " + String(p.weight) + ")")